### PR TITLE
Slightly less crappy Nightmode

### PR DESCRIPTION
**Pull request by NiLuJe**
Companion PR to https://github.com/koreader/koreader-base/pull/884
* Basically flags devices known to be stable when using PxP inversion.
* Plus, random fix for #4870 ;).
* A few FrontLight tweaks & cleanups on Kobo:
  * Moved the Kobo-specific startup status insanity to Kobo-specific init
  * Made turnOff/turnOn frontlight do a smooth ramp down/up
  * On Kobo, use turnOff/turnOn for suspend/resume, to get that smooth toggle
  * On Kobo, for NaturalLight w/ a mixer, only set warmth for setWarmth, and only set Brightness for setBrightness, otherwise, it tried to set both with not in-sync values, which made the FL widget jittery.

### CODE REVIEW

File: frontend/apps/reader/modules/readergesture.lua [hunk 1]
```diff
@@ -703,7 +703,7 @@ function ReaderGesture:gestureAction(action, ges)
     elseif action == "toc" then
         self.ui:handleEvent(Event:new("ShowToc"))
     elseif action == "night_mode" then
-        local night_mode = G_reader_settings:readSetting("night_mode") or false
+        local night_mode = G_reader_settings:isTrue("night_mode") or false
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readergesture.lua]: Redundant or false? ;-)

Comment (NiLuJe) [reply frontend/apps/reader/modules/readergesture.lua]: Indeed, thanks ;).

**Comment by NiLuJe** [comment]
Okay, added some frontlight tweaks on Kobo, because it's kind of a mess there :D. c.f., the commit message from 5b70446 (although it's only worth reviewing the final state, as, err, things moved a lot after testing on various devices :D).

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 2]
```diff
@@ -764,7 +764,7 @@ function ReaderHighlight:onHoldRelease()
                         text = "Copy",
                         enabled = Device:hasClipboard(),
                         callback = function()
-                            Device.input.setClipboardText(self.selected_text.text)
+                            Device.input:setClipboardText(self.selected_text.text)
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerhighlight.lua]: The clipboard functions are actually functions and not methods, so this one (with parameter) should not work anymore as intended when switched from `.` to `:`. The others won't mind, but I guess they should still use `.`. (cf https://github.com/koreader/koreader/pull/4482#issuecomment-454492160 where I thought about making them methods, but didn't :)

**Comment by poire-z** [comment]
(You're courageous for having a take at that frontlight stuff :) I swear I wouldn't anymore after #3143 and #3163, too fragile ... But great if you make that cleaner :)

**Comment by Frenzie** [comment]
Looks alright at a glance; didn't test; nothing to add to @poire-z :-)

File: frontend/apps/reader/modules/readerhighlight.lua (hunk 2 above)
Comment (NiLuJe) [reply frontend/apps/reader/modules/readerhighlight.lua]: By "others" you mean Clipboard stuff, right? Basically, that only leaves hasClipboard`:` ;).

**Comment by NiLuJe** [comment]
I mainly want to murder the unit test that's (failing) to test a non-existent scenario, but the actual changes look sound ;p.

File: frontend/apps/reader/modules/readerhighlight.lua (hunk 2 above)
Comment (poire-z) [reply frontend/apps/reader/modules/readerhighlight.lua]: Yep, (has|get|set)Clipboard(), that you restored as they were it seems, as they are no more in the diff.
Haven't looked at each of the Device:xxx if they are really a method, but mostly everything should be a method :)

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerhighlight.lua]: Yeah, I only rg'd for `evice\.`, so I shouldn't have touched anything else ;).

**Comment by poire-z** [comment]
I'll test all that on my Kobo this evening.
(Looks like 2019.04 is nearly ready for release, so may be wait for it to happen before merging this risky frontlight stuff and new'y HW invert :)

**Comment by Frenzie** [comment]
Yes, I had some time yesterday so I prepped a little release note. I can tag it tomorrow unless there are any urgent remaining issues I'm not aware of?

**Comment by poire-z** [comment]
> unless there are any urgent remaining issues I'm not aware of?

Nope. No real changes since a few days, should be pretty stable :)

**Comment by poire-z** [comment]
> I'll test all that on my Kobo this evening.

No problem noticed so far on my GloHD.
No change noticed with nightmode, but I never really used it.
The progressive off<>on frontlight toggling is nice too (it's smooth when ON>OFF, but when OFF>ON, there is somehow a small pause at ~20% and it feels a bit less smooth - GloHD only?).

**Comment by NiLuJe** [comment]
@poire-z : Noticed that on the H2O, too, so, that'd be all devices using the ioctl (i.e., without NaturalLight).

I'm not sure if it's a quirk of the ioctl itself, or load-related, since it tends to hang just to display the "Frontlight Off" toast.
Which, incidentally, makes it readable, so I'm not *too* bothered by it, but it is weird ;).

**Comment by NiLuJe** [comment]
@poire-z: Yep, it's a hardware quirk.

```C
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/stat.h>

#define CM_FRONT_LIGHT_SET 241

int main(int argc, char **argv)
{
	int ntx_io, retval, ptr;
	ntx_io = open("/dev/ntx_io", O_RDONLY | O_NONBLOCK | O_CLOEXEC);

	uint8_t fl_min = 0U;
	uint8_t fl_max = 100U;
	uint8_t fl_cur = 20U;
	// Ramp-down
	for (uint8_t i = 1U; i <= 5; i++) {
		ptr = __builtin_ifloorf(fl_cur - ((fl_cur / 5.0f) * i));
		retval = ioctl(ntx_io, CM_FRONT_LIGHT_SET, ptr);
		printf("ioctl(ntx_io, CM_FRONT_LIGHT_SET) = %d; returned %d\n", retval, ptr);
		if (i < 5) {
			usleep(35 * 1000);
		}
	}

	usleep(1000 * 1000);
	// Ramp-up
	for (uint8_t i = 1U; i <= 5; i++) {
		ptr = __builtin_iceilf(fl_min + ((fl_cur / 5.0f) * i));
		retval = ioctl(ntx_io, CM_FRONT_LIGHT_SET, ptr);
		printf("ioctl(ntx_io, CM_FRONT_LIGHT_SET) = %d; returned %d\n", retval, ptr);
		if (i < 5) {
			usleep(35 * 1000);
		}
	}

	close(ntx_io);
}
```

Behaves the same way ;).

**Comment by NiLuJe** [comment]
@poire-z : Re-reading your message, you're seeing the stutter on the opposite side I do. Huh. (I'm seeing it when shutting the light off, i.e., the ramp down).

**Pull request merged**